pkg/cmd/pulumi/search.go

```diff
@@ -0,0 +1,163 @@
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/pulumi/pulumi/pkg/v3/backend"
+	"github.com/pulumi/pulumi/pkg/v3/backend/display"
+	"github.com/pulumi/pulumi/pkg/v3/backend/httpstate"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/env"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+	"github.com/spf13/cobra"
+	auto_table "go.pennock.tech/tabular/auto"
+	"go.uber.org/multierr"
+)
+
+type searchCmd struct {
+	orgName     string
+	queryParams []string
+
+	Stdout io.Writer // defaults to os.Stdout
+
+	// currentBackend is a reference to the top-level currentBackend function.
+	// This is used to override the default implementation for testing purposes.
+	currentBackend func(context.Context, *workspace.Project, display.Options) (backend.Backend, error)
+}
+
+func (cmd *searchCmd) Run(ctx context.Context, args []string) error {
+	interactive := cmdutil.Interactive()
+
+	if cmd.Stdout == nil {
+		cmd.Stdout = os.Stdout
+	}
+
+	if cmd.currentBackend == nil {
+		cmd.currentBackend = currentBackend
+	}
+	currentBackend := cmd.currentBackend // shadow the top-level function
+
+	opts := backend.QueryOptions{}
+	opts.Display = display.Options{
+		Color:         cmdutil.GetGlobalColorization(),
+		IsInteractive: interactive,
+		Type:          display.DisplayQuery,
+	}
+	// Try to read the current project
+	project, _, err := readProject()
+	if err != nil && !errors.Is(err, workspace.ErrProjectNotFound) {
+		return err
+	}
+
+	backend, err := currentBackend(ctx, project, opts.Display)
+	if err != nil {
+		return err
+	}
+	cloudBackend, isCloud := backend.(httpstate.Backend)
+	if !isCloud {
+		return errors.New("Pulumi AI search is only supported for the Pulumi Cloud")
+	}
+	userName, orgs, err := cloudBackend.CurrentUser()
+	if err != nil {
+		return err
+	}
+	filterName := userName
+	if cmd.orgName != "" {
+		if !sliceContains(orgs, cmd.orgName) {
+			return fmt.Errorf("user %s is not a member of organization %s", userName, cmd.orgName)
+		}
+		filterName = cmd.orgName
+	}
+
+	parsedQueryParams := apitype.ParseQueryParams(cmd.queryParams)
+	res, err := cloudBackend.Search(ctx, filterName, parsedQueryParams)
+	if err != nil {
+		return err
+	}
+	err = cmd.RenderTable(res.Resources)
+	if err != nil {
+		return fmt.Errorf("table rendering error: %s", err)
+	}
+	return nil
+}
+
+func newSearchCmd() *cobra.Command {
+	var scmd searchCmd
+	cmd := &cobra.Command{
+		Use:   "search",
+		Short: "Search for resources in Pulumi Cloud",
+		Long:  "Search for resources in Pulumi Cloud.",
+		Args:  cmdutil.NoArgs,
+		Run: cmdutil.RunFunc(func(cmd *cobra.Command, args []string) error {
+			ctx := commandContext()
+			return scmd.Run(ctx, args)
+		},
+		),
+	}
+
+	// TODO: Remove this branch once we release this feature fully.
+	if env.Dev.Value() {
+		cmd.AddCommand(newSearchAICmd())
+	}
```

Comment:
Or do we expect to marinate this in dev for a while?

Reply:
I think it's in marination stage at the moment. I'm open to discussion, though.